src/main/kotlin/org/rust/ide/intentions/NestUseStatementsIntention.kt

```diff
@@ -0,0 +1,127 @@
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.RsPsiFactory
+import org.rust.lang.core.psi.RsUseItem
+import org.rust.lang.core.psi.RsUseSpeck
+import org.rust.lang.core.psi.ext.*
+
+class NestUseStatementsIntention : RsElementBaseIntentionAction<NestUseStatementsIntention.Context>() {
+    override fun getText() = "Nest use statements"
+    override fun getFamilyName() = text
+
+    class Context(
+        val useSpecks: List<RsUseSpeck>,
+        val firstUseItem: RsUseItem,
+        val useItems: List<RsUseItem>
+    )
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val useItemList = mutableListOf<RsUseItem>()
+        val useItemOnCursor = element.ancestorStrict<RsUseItem>() ?: return null
+        var firstUseItem = useItemOnCursor
+        useItemList.add(useItemOnCursor)
+
+        var useItem = useItemOnCursor
+        // Collect useItems over cursor
+        while (true) {
+            useItem = useItem.getNextNonCommentSibling() as? RsUseItem ?: break
+            useItemList.add(useItem)
+        }
+
+        useItem = useItemOnCursor
+        // Collect useItems under cursor
+        while (true) {
+            useItem = useItem.getPrevNonCommentSibling() as? RsUseItem ?: break
+            firstUseItem = useItem
+            useItemList.add(useItem)
+        }
+
+        val useSpeckList = useItemList
+            .filter { it.useSpeck != null }
+            .map { it.useSpeck!! }
+        return NestUseStatementsIntention.Context(useSpeckList.toList(), firstUseItem, useItemList)
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val useSpecksByBasePath = mapSpecksByBasePath(ctx)
+
+        val createdUseItems = createGroupedUseItems(useSpecksByBasePath, project)
+
+        val root = ctx.firstUseItem.parent
+        var firstUseSpeck: PsiElement? = null
+        var lastUseSpeck: PsiElement? = null
+
+        for (createdUseItem in createdUseItems) {
+            val addPoint = lastUseSpeck ?: ctx.firstUseItem
+            val added = root.addAfter(createdUseItem, addPoint)
+            lastUseSpeck = added
+            firstUseSpeck = firstUseSpeck ?: added
+        }
+
+        for (useItem in ctx.useItems) {
+            useItem.delete()
+        }
+
+        editor.caretModel.moveToOffset(firstUseSpeck!!.textRange.startOffset + "use ".length)
+    }
+
+    private fun mapSpecksByBasePath(ctx: Context): MutableMap<String, MutableList<RsUseSpeck>> {
+        val useSpeckByBasePath = mutableMapOf<String, MutableList<RsUseSpeck>>()
+        for (useSpeck in ctx.useSpecks) {
+            val useSpeckStartsWithGroup = useSpeck.path == null
+            if (useSpeckStartsWithGroup) {
+                useSpeck.forEachLeafSpeck {
+                    useSpeckByBasePath.getOrPut(it.path!!.basePath().text) { mutableListOf() }.add(it)
```

Comment:
The path is not accidentally marked as nullable. It's because it can be null =) Let's check it
Also it's better to use `.referenceName` instead of `.text`